src/main/java/com/ibm/cloud/sdk/core/security/TokenRequestBasedAuthenticator.java

```diff
@@ -202,6 +202,31 @@ public void authenticate(Builder builder) {
    */
   public abstract T requestToken();
 
+  /**
+   * Calls the extending class' requestToken implementation in a synchronized way, returning the existing token data
+   * if it's been made valid since this method's initial call.
+   *
+   * @return the token object
+   */
+  private synchronized T synchronizedRequestToken() {
+    if (this.tokenData != null && this.tokenData.isTokenValid()) {
+      return this.tokenData;
+    }
+
+    return requestToken();
+  }
+
+  /**
+   * Requests a new token after performing one last check that the token still needs refreshing, in case another thread
+   * has since refreshed it.
+   */
+  private synchronized void refreshToken() {
+    if (!this.tokenData.needsRefresh()) {
+      return;
+    }
+    this.tokenData = synchronizedRequestToken();
+  }
+
```

Comment:
I don't think you need this function at all -- we should call `requestToken` directly when we need to refresh the token.

Reply:
All these changes make it way cleaner. Thank you!